lib/SILOptimizer/Transforms/DeadCodeElimination.cpp

```diff
@@ -266,6 +265,21 @@ void DCE::markLive() {
         break;
       }
       case SILInstructionKind::BeginBorrowInst: {
+        // Currently we only support borrows of owned values.
+        // Nested borrow handling can be complex in the presence of reborrows.
+        // So it is not handled currently.
+        auto *borrowInst = cast<BeginBorrowInst>(&I);
+        if (borrowInst->getOperand().getOwnershipKind() !=
+            OwnershipKind::Owned) {
+          markInstructionLive(borrowInst);
+          // Visit all end_borrows and mark them live
+          auto visitEndBorrow = [&](EndBorrowInst *endBorrow) {
+            markInstructionLive(endBorrow);
+          };
+          visitTransitiveEndBorrows(borrowInst, visitEndBorrow);
```

Comment:
Is visiting the end-borrows necessary? Wouldn't be a nested end_borrow be kept alive by the marking begin_borrows with borrowed operands alive (5 lines above) + the reverse dependency?

Reply:
I added this to specifically handle the reborrow case, where a borrowed value is passed as a phi argument. And the end_borrow is on the phi argument. There is no reverse dependency between the end_borrow and the initial begin_borrow in this case. The test `dce_nestedborrowlifetime3` in `dead_code_elimination_nontrivial_ossa.sil` shows an example.

Reply:
ok